From 471857710f2ee17500413548e3229a2108c31584 Mon Sep 17 00:00:00 2001
From: Michal Kubecek <mkubecek@suse.cz>
Date: Mon, 11 Mar 2019 08:07:38 +0100
Subject: [PATCH 32/44] ethtool: set pause parameters with SET_PARAMS request

Add support for setting device pause parameters using SET_SETTINGS request
with ETHA_PARAMS_PAUSE nested attribute. This is a replacement of
ETHTOOL_SPAUSEPARAM ioctl command.

Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
---
 Documentation/networking/ethtool-netlink.txt | 14 +++---
 net/ethtool/params.c                         | 47 +++++++++++++++++++-
 2 files changed, 55 insertions(+), 6 deletions(-)

diff --git a/Documentation/networking/ethtool-netlink.txt b/Documentation/networking/ethtool-netlink.txt
index a31883040474..f29c8c3a2fc7 100644
--- a/Documentation/networking/ethtool-netlink.txt
+++ b/Documentation/networking/ethtool-netlink.txt
@@ -477,10 +477,10 @@ SET_PARAMS
 ----------
 
 SET_PARAMS request modifies the settings retrieved by GET_PARAMS, i.e. it
-replaces ETHTOOL_SCOALESCE and ETHTOOL_SRINGPARAM ioctl commands. For each of
-these, relevant data attributes are contained in a corresponding nest
-attribute. Some of the attributes provided by GET_SETPARAMS are read only and
-cannot be set by SET_PARAMS request.
+replaces ETHTOOL_SCOALESCE, ETHTOOL_SRINGPARAM and ETHTOOL_SPAUSEPARAM ioctl
+commands. For each of these, relevant data attributes are contained in
+a corresponding nest attribute. Some of the attributes provided by
+GET_SETPARAMS are read only and cannot be set by SET_PARAMS request.
 
     ETHA_PARAMS_COALESCE	(nested)	coalescing parameters
         ETHA_COALESCE_RX_USECS			(u32)
@@ -510,6 +510,10 @@ cannot be set by SET_PARAMS request.
         ETHA_RING_RX_MINI_PENDING		(u32)
         ETHA_RING_RX_JUMBO_PENDING		(u32)
         ETHA_RING_TX_PENDING			(u32)
+    ETHA_PARAMS_PAUSE		(nested)	pause parameters
+        ETHA_PAUSE_AUTONEG			(bool)
+        ETHA_PAUSE_RX				(bool)
+        ETHA_PAUSE_TX				(bool)
 
 
 Request translation
@@ -538,7 +542,7 @@ ETHTOOL_SCOALESCE		ETHNL_CMD_SET_PARAMS
 ETHTOOL_GRINGPARAM		ETHNL_CMD_GET_PARAMS
 ETHTOOL_SRINGPARAM		ETHNL_CMD_SET_PARAMS
 ETHTOOL_GPAUSEPARAM		ETHNL_CMD_GET_PARAMS
-ETHTOOL_SPAUSEPARAM		n/a
+ETHTOOL_SPAUSEPARAM		ETHNL_CMD_SET_PARAMS
 ETHTOOL_GRXCSUM			ETHNL_CMD_GET_SETTINGS
 ETHTOOL_SRXCSUM			ETHNL_CMD_SET_SETTINGS
 ETHTOOL_GTXCSUM			ETHNL_CMD_GET_SETTINGS
diff --git a/net/ethtool/params.c b/net/ethtool/params.c
index bae4398ced74..ff4999fc279c 100644
--- a/net/ethtool/params.c
+++ b/net/ethtool/params.c
@@ -300,7 +300,7 @@ static const struct nla_policy set_params_policy[ETHA_PARAMS_MAX + 1] = {
 	[ETHA_PARAMS_COMPACT]		= { .type = NLA_FLAG },
 	[ETHA_PARAMS_COALESCE]		= { .type = NLA_NESTED },
 	[ETHA_PARAMS_RING]		= { .type = NLA_NESTED },
-	[ETHA_PARAMS_PAUSE]		= { .type = NLA_REJECT },
+	[ETHA_PARAMS_PAUSE]		= { .type = NLA_NESTED },
 };
 
 static const struct nla_policy coalesce_policy[ETHA_COALESCE_MAX + 1] = {
@@ -489,6 +489,46 @@ static int update_ring(struct genl_info *info, struct net_device *dev,
 	return (ret < 0) ? ret : 1;
 }
 
+static const struct nla_policy pause_policy[ETHA_PAUSE_MAX + 1] = {
+	[ETHA_PAUSE_UNSPEC]	= { .type = NLA_REJECT },
+	[ETHA_PAUSE_AUTONEG]	= { .type = NLA_U8 },
+	[ETHA_PAUSE_RX]		= { .type = NLA_U8 },
+	[ETHA_PAUSE_TX]		= { .type = NLA_U8 },
+};
+
+static int update_pause(struct genl_info *info, struct net_device *dev,
+			struct nlattr *nest)
+{
+	struct nlattr *tb[ETHA_RING_MAX + 1];
+	struct ethtool_pauseparam data = {};
+	bool mod = false;
+	int ret;
+
+	if (!nest)
+		return 0;
+	if (!dev->ethtool_ops->get_pauseparam ||
+	    !dev->ethtool_ops->set_pauseparam)
+		return -EOPNOTSUPP;
+	dev->ethtool_ops->get_pauseparam(dev, &data);
+
+	ret = nla_parse_nested_strict(tb, ETHA_PAUSE_MAX, nest, pause_policy,
+				      info->extack);
+	if (ret < 0)
+		return ret;
+
+	if (ethnl_update_u32(&data.autoneg, tb[ETHA_PAUSE_AUTONEG]))
+		mod = true;
+	if (ethnl_update_u32(&data.rx_pause, tb[ETHA_PAUSE_RX]))
+		mod = true;
+	if (ethnl_update_u32(&data.tx_pause, tb[ETHA_PAUSE_TX]))
+		mod = true;
+
+	if (!mod)
+		return 0;
+	ret = dev->ethtool_ops->set_pauseparam(dev, &data);
+	return (ret < 0) ? ret : 1;
+}
+
 int ethnl_set_params(struct sk_buff *skb, struct genl_info *info)
 {
 	struct nlattr *tb[ETHA_PARAMS_MAX + 1];
@@ -518,6 +558,11 @@ int ethnl_set_params(struct sk_buff *skb, struct genl_info *info)
 		goto out_ops;
 	if (ret)
 		req_mask |= ETH_PARAMS_IM_RING;
+	ret = update_pause(info, dev, tb[ETHA_PARAMS_PAUSE]);
+	if (ret < 0)
+		goto out_ops;
+	if (ret)
+		req_mask |= ETH_PARAMS_IM_PAUSE;
 
 	ret = 0;
 out_ops:
-- 
2.21.0

